# LAB 1 - document classification

**Ques 1 :** Write a program to read a file and convert it into a vector of unique words based on term frequency

In [5]:
import string

def file_to_term_frequency_vector(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        text = f.read()

    text = text.lower()
    text = text.translate(str.maketrans('', '', string.punctuation)) 
    words = text.split()

    stopwords = {'a', 'an', 'the', 'of', 'for', 'has', 'have', 'been', 'in', 'on', 'to', 
                 'this', 'that', 'from', 'as', 'is', 'are', 'and', 'at', 'do', 'how', 'what', 'where', 'why', 'which'}

    freq_dict = {}
    for word in words:
        if word in stopwords:
            continue
        if word in freq_dict:
            freq_dict[word] += 1
        else:
            freq_dict[word] = 1

    sorted_freq = dict(sorted(freq_dict.items(), key=lambda x: x[1], reverse=True)) # soert by highest freq
    return sorted_freq

file_path = input("ENTER FILE NAME: ")
freq_vector = file_to_term_frequency_vector(file_path)

for word, freq in freq_vector.items():
    print(f'{word}: {freq}')

ENTER FILE NAME:  document001.txt


computing: 7
computers: 6
power: 4
evolution: 2
vacuum: 2
machines: 2
quantum: 2
development: 2
could: 2
patterns: 2
systems: 2
simulating: 2
fascinating: 1
journey: 1
radically: 1
transformed: 1
society: 1
early: 1
days: 1
tubebased: 1
today’s: 1
cuttingedge: 1
technologies: 1
nothing: 1
short: 1
revolutionary: 1
initially: 1
were: 1
large: 1
roomsized: 1
perform: 1
only: 1
basic: 1
arithmetic: 1
however: 1
transistors: 1
replaced: 1
tubes: 1
became: 1
smaller: 1
faster: 1
more: 1
reliable: 1
advent: 1
microprocessors: 1
1970s: 1
marked: 1
beginning: 1
personal: 1
computer: 1
era: 1
bringing: 1
homes: 1
small: 1
businesses: 1
shift: 1
paved: 1
way: 1
software: 1
innovations: 1
rise: 1
internet: 1
recent: 1
years: 1
advancements: 1
artificial: 1
intelligence: 1
ai: 1
machine: 1
learning: 1
driving: 1
next: 1
wave: 1
today: 1
can: 1
analyze: 1
vast: 1
amounts: 1
data: 1
learn: 1
make: 1
decisions: 1
autonomously: 1
highperformance: 1
hpc: 1
now: 1
capable: 1
complex: 1
weather: 1
molecu

**Ques 2 :** Write a program to read multiple files and make a vector by taking 10 most frequent words for each document and make a vector of the frequent words in all n documents

In [44]:
import string
import math

def file_to_term_frequency_vector(file_path):
    with open(file_path, 'r', encoding='utf-8') as f:
        text = f.read()

    text = text.lower()
    text = text.translate(str.maketrans('', '', string.punctuation))
    words = text.split()

    stopwords = {'a', 'an', 'the', 'of', 'for', 'has', 'have', 'in', 'on', 'to', 
             'this', 'that', 'from', 'as', 'is', 'are'}

    freq_dict = {}
    for word in words:
        if word in stopwords:
            continue
        if word in freq_dict:
            freq_dict[word] += 1
        else:
            freq_dict[word] = 1

    sorted_freq = dict(sorted(freq_dict.items(), key=lambda x: x[1], reverse=True))
    return sorted_freq

def get_top_words(freq_dict, top_n=10):
    return list(freq_dict.keys())[:top_n]

def build_vector(freq_dict, vocabulary):
    return [freq_dict.get(word, 0) for word in vocabulary] # taking reference from vocab, find the word's frequency, if not found set to 0

def cosine_similarity(vec1, vec2):
    dot = sum(a * b for a, b in zip(vec1, vec2)) # zipped the vectors(lists) together to combine ele at same posn -> then multiplied correspondingly
    norm1 = math.sqrt(sum(a * a for a in vec1)) # magnitude -- easy
    norm2 = math.sqrt(sum(b * b for b in vec2))
    if norm1 == 0 or norm2 == 0:
        return 0.0
    return dot / (norm1 * norm2)

# Action begins here

file_paths = ['document001.txt', 'document002.txt', 'document003.txt']

freq_dicts = [file_to_term_frequency_vector(path) for path in file_paths]

top_words = []
for freq in freq_dicts:
    top_words.extend(get_top_words(freq, top_n=10))

vocabulary = sorted(set(top_words)) # converted to set to remove duplicates

vectors = [build_vector(freq, vocabulary) for freq in freq_dicts]

n = len(file_paths)
print("\nCosine Similarity Matrix:\n")
print("   |---D1---|---D2---|---D3---|")
for i in range(n):
    print(f"D{i+1}", end='  ')
    for j in range(n):
        sim = cosine_similarity(vectors[i], vectors[j])
        print(f"{sim:.4f}", end='    ')
    print()


Cosine Similarity Matrix:

   |---D1---|---D2---|---D3---|
D1  1.0000    0.6575    0.4952    
D2  0.6575    1.0000    0.5312    
D3  0.4952    0.5312    1.0000    


# Similarity Result
- Doc1 and Doc2 : 65%
- Doc1 and Doc3 : 49%
- Doc2 and Doc3 : 53%